### PR TITLE
Return authentication context for failed Attempts

### PR DESCRIPTION
**Pull request by IsuraD**

### CODE REVIEW

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java [hunk 1]
```diff
@@ -1637,42 +1638,23 @@ public AuthenticationContext authenticate(Claim claim, Callback[] credentials, S
...
-                    String uniqueUserId;
-                    try {
-                        uniqueUserId = getEncodedUniqueEntityId(domain.getId(), domainUserId);
-                    } catch (IdentityStoreException e) {
-                        throw new IdentityStoreServerException("Failed to build unique user id.");
+                    AuthenticationContext authenticationContext = domain.authenticate(claim, credentials);
+                    if (authenticationContext.isAuthenticated()) {
+                        return getAuthenticationContext(authenticationContext, domain);
+                    } else {
+                        DomainUser domainUser = (DomainUser) authenticationContext.getParameter(
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java]: Do we need to keep the domain user in the property map after using it?

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java [hunk 2]
```diff
@@ -1739,6 +1721,35 @@ private String getEncodedUniqueEntityId(int domainId, String domainEntityId) thr
     }
 
     /**
+     * Get AuthenticationContext from DomainAuthenticationContext
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java]: DomainAuthenticationContext is no longer valid.

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java [hunk 3]
```diff
@@ -1739,6 +1721,35 @@ private String getEncodedUniqueEntityId(int domainId, String domainEntityId) thr
...
+     *
+     * @param authenticationContext : domainAuthenticationContext
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java]: Missing method comment annotations. @param domain and @throws

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java [hunk 4]
```diff
@@ -1739,6 +1721,35 @@ private String getEncodedUniqueEntityId(int domainId, String domainEntityId) thr
...
+                    .setState(domainUser.getState())
+                    .setDomainName(domain.getName()).build());
+            return authenticationContext;
+        } else {
+            if (authenticationContext instanceof FailedAuthenticationContext) {
+                ((FailedAuthenticationContext) authenticationContext).addFailedUserIdToList(getEncodedUniqueEntityId(
+                        domain.getId(), domainUser.getDomainUserId()));
+                return authenticationContext;
+            }
+            throw new IdentityStoreException("Invalid Authentication context ");
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java]: Remove extra space after `context`

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java (hunk 1 above)
Comment (IsuraD) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/impl/IdentityStoreImpl.java]: Better to keep that. No need to remove

**Pull request merged**